oc_ldap_user_locker/mailer.py

```diff
@@ -0,0 +1,151 @@
+import smtplib
+import urllib.parse as urlparse
+import re
+import posixpath
+
+class LockMailer:
+    def __init__(self, config, config_path):
+        """
+        Basic initialization, configuration checking
+        :param dict config: configuration for mailer
+        :param str base_path: path to a directory with basic configuration
+        """
+        self._config = config
+        self._config_path = os.path.abspath(config_path)
+        logging.debug("Base configutaion path: '%s'" % self._config_path)
+        self._check_config()
+
+    def _check_config(self):
+        """
+        Check mailer configuration and adjust paths if necessary
+        """
+
+        _smtp_env = {
+                "url": None,
+                "user": None,
+                "password": None,
+                "from": "MAIL_FROM"
+                }
+
+        if not self._config:
+            logging.debug("SMTP configuration missing, trying to create it from environment")
+            self._config = dict()
+
+        for _key in _smtp_env.keys():
+            _env = _smtp_env.get(_key) or '_'.join(["SMTP", _key.upper()])
+            _value = self._config.get(_key, os.getenv(_env))
+
+            if _value:
+                logging.debug("%s: '%s'" % (_env, _value))
+                self._config[_key] = _value
+
+        # 'url' and 'from' are required, 'user' and 'password' may be omitted for anonymous SMTP
+        for _k in ["url", "from"]:
+            if not self._config.get(_k):
+                raise ValueError("Required '%s' is not set for SMTP" % _k)
+
+    def _check_path(self, path):
+        """
+        Check if path provided is absolute.
+        Make it absolute if not.
+        Do not check for file existence since it will be raised while try to open
+        :param str path: path to check
+        :return str: adjusted path
+        """
+        if not path:
+            raise ValueError("Empty or incompatible path provided")
+
+        if not os.path.isabs(path):
+            path = os.path.join(self._config_path, path)
+
+        return path
+
+    def _check_template_configuration(self, template_conf):
+        """
+        Check template configuration
+        :param dict template_conf: template configuration
+        :return dict: adjusted template configuration
+        """
+        if not isinstance(template_conf, dict):
+            raise TypeError("Dictinary required, %s provided" % type(template_conf))
+
+        template_conf["file"] = self._check_path(template_conf["file"])
+        template_conf["type"] = template_conf.get("type") or "plain"
+
+        if "signature" in template_conf.keys():
+            template_conf["signature"] = self._check_path(template_conf["signature"])
+
+        return template_conf
+
+    def _get_smtp_client(self):
+        """
+        Return SMTP connection instance
+        """
+        _url = self._config.get("url")
+
+        if not re.match("(.*?:)?" + posixpath.sep + posixpath.sep, _url):
```

Comment:
Maybe urllib.parse.urlpase.scheme?

Reply:
Unfortunately not.
This trick is implemented for `urlparse` to work correctly.
Example:
```
>>> urlparse.urlparse("bobby:1525/robert").scheme
'bobby'
```